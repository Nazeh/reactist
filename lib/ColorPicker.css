.reactist.color_picker .color_trigger {
  flex: 0 0 auto;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 24px;
  width: 24px;
  border-radius: 50%;
  margin: 5px;
  cursor: pointer;
  background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0Ij4gICAgPHBhdGggZmlsbD0iI0ZGRiIgZmlsbC1ydWxlPSJldmVub2RkIiBkPSJNOS4yNSAxMC42NDZsMi42NDYgMi42NDcgMi42NDctMi42NDdhLjUuNSAwIDAgMSAuNzA3LjcwOEwxMi42MDQgMTRhMSAxIDAgMCAxLTEuNDE1IDBsLTIuNjQ2LTIuNjQ2YS41LjUgMCAwIDEgLjcwNy0uNzA4eiIvPjwvc3ZnPg==);
  background-position: center;
  background-repeat: no-repeat;
  background-size: 24px;
}
.reactist.color_picker .color_trigger--inner_ring {
  background-color: rgba(0, 0, 0, 0.1);
  width: 14px;
  height: 14px;
  border-radius: 50%;
  visibility: hidden;
}
.reactist.color_picker .color_trigger:hover .color_trigger--inner_ring {
  visibility: visible;
}
.reactist.color_picker .color_item {
  flex: 0 0 auto;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 4px;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  cursor: pointer;
}
.reactist.color_picker .color_item--inner_ring {
  background-color: transparent;
  border: 2px solid white;
  height: 24px;
  width: 24px;
  border-radius: 50%;
  visibility: hidden;
}
.reactist.color_picker .color_item:hover .color_item--inner_ring {
  visibility: visible;
}
.reactist.color_picker .color_item.active {
  background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0Ij4gICAgPHBhdGggZmlsbD0iI0ZGRiIgZmlsbC1ydWxlPSJldmVub2RkIiBkPSJNMTAuOTc0IDE0Ljc3MWw0LjE2LTcuMjA0YS40OTkuNDk5IDAgMSAxIC44NjMuNDk5bC00LjQ3IDcuNzQ0YS40OTkuNDk5IDAgMCAxLS43MzUuMTQ3LjUwMi41MDIgMCAwIDEtLjA2MS0uMDQ4bC0zLjM2Ni0zLjE0YS40OTkuNDk5IDAgMSAxIC42OC0uNzI5bDIuOTI5IDIuNzMxeiIvPjwvc3ZnPg==);
  background-position: center;
  background-repeat: no-repeat;
}
.reactist.color_picker .color_options {
  padding: 6px;
  display: flex;
  flex-wrap: wrap;
  width: 164px;
  position: relative;
}

.reactist.dropdown .trigger {
  cursor: pointer;
}
.reactist.dropdown .body {
  border-radius: 3px;
  border: 1px solid #dcdcdc;
  overflow: hidden;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
  z-index: 1;
  background-color: #fff;
}
.reactist.dropdown hr {
  border: none;
  height: 1px;
  background-color: #ececec;
  margin: 0 5px;
}
.reactist.dropdown .with_arrow:after,
.reactist.dropdown .with_arrow:before {
  z-index: 1;
  content: '';
  display: block;
  position: absolute;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 6px;
  right: 14px;
}
.reactist.dropdown .with_arrow:after {
  top: -11px;
  border-color: transparent transparent #fff transparent;
}
.reactist.dropdown .with_arrow:before {
  top: -12px;
  border-color: transparent transparent #dcdcdc transparent;
}
.reactist.dropdown .with_arrow.top:after {
  top: -1px;
  border-color: #fff transparent transparent transparent;
}
.reactist.dropdown .with_arrow.top:before {
  top: -1px;
  right: 13px;
  border-width: 7px;
  border-color: #dcdcdc transparent transparent transparent;
}

.reactist.tooltip {
  background-color: #333333;
  border-radius: 3px;
  padding: 5px 10px;
  max-width: 250px;
  white-space: normal;
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  visibility: hidden;
  opacity: 0;
  transition: opacity 0.1s linear;
}
.reactist.tooltip__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist.tooltip .tooltip__text {
  font-size: 0.81rem;
  color: white;
  font-weight: 400;
  line-height: 1.6;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  display: inline-block;
}
.reactist.tooltip.visible {
  visibility: visible;
  opacity: 1;
}
.reactist.tooltip.inverted {
  background-color: white;
  border-radius: 3px;
  border: 1px solid #dcdcdc;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
}
.reactist.tooltip.inverted .tooltip__text {
  color: #202020;
}
.reactist.tooltip.arrow_top:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #333333;
}
.reactist.tooltip.arrow_top.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #dcdcdc;
  top: calc(-5px - 1px);
}
.reactist.tooltip.arrow_top.inverted:after {
  border-bottom: 5px solid white;
}
.reactist.tooltip.arrow_right:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #333333;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist.tooltip.arrow_right.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
  right: calc(-5px - 1px);
}
.reactist.tooltip.arrow_right.inverted:after {
  border-left: 5px solid white;
}
.reactist.tooltip.arrow_bottom:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #333333;
}
.reactist.tooltip.arrow_bottom.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #dcdcdc;
  bottom: calc(-5px - 1px);
}
.reactist.tooltip.arrow_bottom.inverted:after {
  border-top: 5px solid white;
}
.reactist.tooltip.arrow_left:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #333333;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist.tooltip.arrow_left.inverted:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
  left: calc(-5px - 1px);
}
.reactist.tooltip.arrow_left.inverted:after {
  border-right: 5px solid white;
}


/*# sourceMappingURL=ColorPicker.map*/