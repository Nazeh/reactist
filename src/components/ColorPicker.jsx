import './styles/color_picker.less'

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Dropdown from './Dropdown'
import Tooltip from './Tooltip'

const COLORS = [
    '#606060',
    '#4A90E2',
    '#03B3B2',
    '#008299',
    '#82BA00',
    '#D24726',
    '#AC193D',
    '#DC4FAD',
    '#3BD5FB',
    '#74E8D3',
    '#FFCC00',
    '#FB886E',
    '#CCCCCC'
]

const _isNamedColor = color => typeof color !== 'string'
const _getColor = (colorList, colorIndex) => {
    const index = colorIndex >= colorList.length ? 0 : colorIndex
    return colorList[index]
}

const ColorPicker = ({ color, small, onChange, colorList = COLORS }) => (
    <Dropdown.Box right className="reactist_color_picker">
        <Dropdown.Trigger>
            <span
                className={classnames('color_trigger', { small })}
                style={{
                    backgroundColor: _isNamedColor(colorList[color])
                        ? _getColor(colorList, color).color
                        : _getColor(colorList, color)
                }}
            >
                <span className="color_trigger--inner_ring" />
            </span>
        </Dropdown.Trigger>
        <Dropdown.Body>
            <div className="color_options">
                {colorList.reduce((items, currentColor, currentIndex) => {
                    const isNamed = _isNamedColor(currentColor)
                    items.push(
                        <ColorItem
                            isActive={
                                color >= colorList.length
                                    ? currentIndex === 0
                                    : currentIndex === color
                            }
                            key={currentIndex}
                            color={isNamed ? currentColor.color : currentColor}
                            colorIndex={currentIndex}
                            onClick={onChange}
                            tooltip={isNamed ? currentColor.name : null}
                        />
                    )
                    return items
                }, [])}
            </div>
        </Dropdown.Body>
    </Dropdown.Box>
)
ColorPicker.displayName = 'ColorPicker'
ColorPicker.defaultProps = {
    color: 0
}
ColorPicker.propTypes = {
    /** Currently selected color. Needs to be the index of the COLORS array. */
    color: PropTypes.number.isRequired,
    /** Callback that is invoked when a color has been selected. Is called with the index of the COLORS array. */
    onChange: PropTypes.func,
    /** Optional flag whether a smaller version of the color picker should be rendered. */
    small: PropTypes.bool,
    /** Optional list of color codes. Either as an array of strings or an array of objects with the color name. Defaults to COLORS array without names. */
    colorList: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.shape({ color: PropTypes.string, name: PropTypes.string })
        ])
    )
}

const ColorItem = ({ color, colorIndex, isActive, onClick, tooltip }) => {
    const item = (
        <span
            className={'reactist color_item' + (isActive ? ' active' : '')}
            style={{ backgroundColor: color }}
            onClick={() => onClick && onClick(colorIndex)}
        >
            <span className="color_item--inner_ring" />
        </span>
    )

    return tooltip ? <Tooltip text={tooltip}>{item}</Tooltip> : item
}
ColorItem.displayName = 'ColorItem'
ColorItem.propTypes = {
    /** The color of the ColorItem as string. */
    color: PropTypes.string.isRequired,
    /** Index of the color to display. Is based upon the colorList array. */
    colorIndex: PropTypes.number.isRequired,
    /** Flag that can be used to highlight the currently selected item. */
    isActive: PropTypes.bool,
    /** Optional callback that is called when the item is clicked. */
    onClick: PropTypes.func,
    /** Optional tooltip to be shown when hovering the item. */
    tooltip: PropTypes.string
}

export default ColorPicker
export { ColorItem, COLORS }
