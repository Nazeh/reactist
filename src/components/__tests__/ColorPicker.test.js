import React from 'react'
import { shallow } from 'enzyme'

import ColorPicker, { ColorItem } from '../ColorPicker'

describe('ColorPicker', () => {
    it('renders without crashing', () => {
        const colorPicker = shallow(getColorPicker())
        expect(colorPicker).toMatchSnapshot()
    })

    it('renders with custom colorList', () => {
        const colorPicker = shallow(
            getColorPicker({ colorList: ['red', 'green', '#0000FF'] })
        )
        expect(colorPicker).toMatchSnapshot()
    })

    it('renders all colors except the selected one', () => {
        const colorPicker = shallow(getColorPicker({ color: 3 }))

        const colorOptions = colorPicker.find('.color_options')
        const selectedColorItem = colorOptions.find('ColorItem[colorIndex=3]')
        expect(selectedColorItem).toHaveLength(0)
    })

    describe('ColorItem', () => {
        it('renders given color and does nothing when clicked without specified onClick handler', () => {
            const colorItem = shallow(getColorItem())
            colorItem.simulate('click')
            expect(colorItem).toMatchSnapshot()
        })

        it('calls onClick after clicking it', () => {
            const onClickSpy = jest.fn()
            const colorItem = shallow(
                getColorItem({ colorIndex: 5, onClick: onClickSpy })
            )
            colorItem.simulate('click')
            expect(onClickSpy).toHaveBeenLastCalledWith(5)
        })
    })

    // Helpers ================================================================
    const getColorPicker = props => <ColorPicker {...props} />
    const getColorItem = props => (
        <ColorItem color="#606060" colorIndex={0} {...props} />
    )
})
