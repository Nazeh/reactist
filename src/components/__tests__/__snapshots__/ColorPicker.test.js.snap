// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker ColorItem renders given color and does nothing when clicked without specified onClick handler 1`] = `
<span
  className="reactist color_item"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#606060",
    }
  }
>
  <span
    className="color_item--inner_ring"
  />
</span>
`;

exports[`ColorPicker renders with custom colorList 1`] = `
<Dropdown.Box
  className="reactist_color_picker"
  right={true}
>
  <Dropdown.Trigger>
    <span
      className="color_trigger"
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    >
      <span
        className="color_trigger--inner_ring"
      />
    </span>
  </Dropdown.Trigger>
  <Dropdown.Body>
    <div
      className="color_options"
    >
      <ColorItem
        color="red"
        colorIndex={0}
        isActive={true}
        key="0"
        tooltip={null}
      />
      <ColorItem
        color="green"
        colorIndex={1}
        isActive={false}
        key="1"
        tooltip={null}
      />
      <ColorItem
        color="#0000FF"
        colorIndex={2}
        isActive={false}
        key="2"
        tooltip={null}
      />
    </div>
  </Dropdown.Body>
</Dropdown.Box>
`;

exports[`ColorPicker renders without crashing 1`] = `
<Dropdown.Box
  className="reactist_color_picker"
  right={true}
>
  <Dropdown.Trigger>
    <span
      className="color_trigger"
      style={
        Object {
          "backgroundColor": "#606060",
        }
      }
    >
      <span
        className="color_trigger--inner_ring"
      />
    </span>
  </Dropdown.Trigger>
  <Dropdown.Body>
    <div
      className="color_options"
    >
      <ColorItem
        color="#606060"
        colorIndex={0}
        isActive={true}
        key="0"
        tooltip={null}
      />
      <ColorItem
        color="#4A90E2"
        colorIndex={1}
        isActive={false}
        key="1"
        tooltip={null}
      />
      <ColorItem
        color="#03B3B2"
        colorIndex={2}
        isActive={false}
        key="2"
        tooltip={null}
      />
      <ColorItem
        color="#008299"
        colorIndex={3}
        isActive={false}
        key="3"
        tooltip={null}
      />
      <ColorItem
        color="#82BA00"
        colorIndex={4}
        isActive={false}
        key="4"
        tooltip={null}
      />
      <ColorItem
        color="#D24726"
        colorIndex={5}
        isActive={false}
        key="5"
        tooltip={null}
      />
      <ColorItem
        color="#AC193D"
        colorIndex={6}
        isActive={false}
        key="6"
        tooltip={null}
      />
      <ColorItem
        color="#DC4FAD"
        colorIndex={7}
        isActive={false}
        key="7"
        tooltip={null}
      />
      <ColorItem
        color="#3BD5FB"
        colorIndex={8}
        isActive={false}
        key="8"
        tooltip={null}
      />
      <ColorItem
        color="#74E8D3"
        colorIndex={9}
        isActive={false}
        key="9"
        tooltip={null}
      />
      <ColorItem
        color="#FFCC00"
        colorIndex={10}
        isActive={false}
        key="10"
        tooltip={null}
      />
      <ColorItem
        color="#FB886E"
        colorIndex={11}
        isActive={false}
        key="11"
        tooltip={null}
      />
      <ColorItem
        color="#CCCCCC"
        colorIndex={12}
        isActive={false}
        key="12"
        tooltip={null}
      />
    </div>
  </Dropdown.Body>
</Dropdown.Box>
`;
